### ISSUE TITLE
[css-images] Images on the web

### ISSUE THREAD

**Issue by annevk** [w3c/csswg-drafts#5173]
I think it's problematic that both CSS and HTML attempt to define images for the web.

We need a central place that is responsible for decoding a byte sequence/stream (and dealing with EXIF) into an abstract representation of an image, that can then be used for various purposes. If we define this in different places in conflicting ways, the differences can lead to security issues.

See also https://github.com/w3c/csswg-drafts/issues/4929 and https://github.com/w3c/csswg-drafts/issues/5165.

cc @domenic

**Comment by annevk** [comment]
@tabatkins @fantasai @LeaVerou thoughts? https://github.com/whatwg/html/pull/5574 wants to standardize more EXIF data. As pointed out above we need a shared strategy for metadata and opaque responses. It would probably be good to make some progress here on these issues to make the fundamentals a bit clearer to implementers, especially now that it's still somewhat early days for taking EXIF data into account.

cc @noamr

**Comment by annevk** [comment]
cc @heycam

**Comment by lrosenthol** [comment]
And as more embedded metadata values will become important to browsers, I would agree that a single place to address the situation would be a good thing.

**Comment by faceless2** [comment]
>  to make the fundamentals a bit clearer to implementers,

👋 

We've walked right into this one - I thought we'd fully implemented css-images-3 but, having found https://github.com/whatwg/html/pull/5574 and the corresponding testcases I can see that's the wrong thing to do if we want to match browser implementations.

As specified in css-images: the image-resolution (theoretically from the `image-resolution` property, although it's unimplemented) defaults to 1dppx, but can be set to `from-image` to take the _natural resolution_ from a bitmap. If the image is loaded from `image-set()`, the natural resolution is _overridden_ by the resolution from there.  The exact definition of _natural resolution_ is undefined in CSS, but given it is defined in JPEG (JFIF) and PNG (pHYs) it seemed clear enough.

What actually happens is the image resolution defaults to the natural resolution of the image, but only if its specified by a particular combination of EXIF tags rather than the tags designed for this purpose in the JPEG and PNG specs (for compat reasons - having read the full PR above I understand why, but this isn't even hinted at in css-images-3 and it's far from obvious). If loaded from `image-set()` the natural resolution is _multiplied_ by the resolution from there (see https://github.com/w3c/csswg-drafts/issues/6241)

It's quite different. Particular issues:

1. The rules in the HTML spec apply even to `background-image`, `content`, SVG `<image>` and contexts where I had not expected the HTML image sizing rules to apply.
2. The rules for extracting natural dimensions are very specific and cannot be applied to PNG, despite PNG images with resolution units being used for some related examples, eg https://webkit.org/demos/srcset/. These are _false friends_ - the resolution units are in the files, the example is demonstrating resolution units, but those facts are unrelated: the units from the file are not used. This only became clear when we revisited [#6241](https://github.com/w3c/csswg-drafts/issues/6241) and found it meant scaling was now being applied twice.

I write this in the hope an anatomy of our failure can improve things, but the short version is:

> I think it's problematic that both CSS and HTML attempt to define images for the web.

**Comment by noamr** [comment]
> > to make the fundamentals a bit clearer to implementers,
> 
> 👋
> 
> We've walked right into this one - I thought we'd fully implemented css-images-3 but, having found [whatwg/html#5574](https://github.com/whatwg/html/pull/5574) and the corresponding testcases I can see that's the wrong thing to do if we want to match browser implementations.
> 
> As specified in css-images: the image-resolution (theoretically from the `image-resolution` property, although it's unimplemented) defaults to 1dppx, but can be set to `from-image` to take the _natural resolution_ from a bitmap. If the image is loaded from `image-set()`, the natural resolution is _overridden_ by the resolution from there. The exact definition of _natural resolution_ is undefined in CSS, but given it is defined in JPEG (JFIF) and PNG (pHYs) it seemed clear enough.
> 
> What actually happens is the image resolution defaults to the natural resolution of the image, but only if its specified by a particular combination of EXIF tags rather than the tags designed for this purpose in the JPEG and PNG specs (for compat reasons - having read the full PR above I understand why, but this isn't even hinted at in css-images-3 and it's far from obvious). If loaded from `image-set()` the natural resolution is _multiplied_ by the resolution from there (see #6241)
> 
> It's quite different. Particular issues:
> 
> 1. The rules in the HTML spec apply even to `background-image`, `content`, SVG `<image>` and contexts where I had not expected the HTML image sizing rules to apply.

HTML sizing rules have always applied to CSS - e.g. determining the default tile size in backgrounds. I am surprised though that there is no explicit link from [CSS natural sizing definition](https://www.w3.org/TR/css-images-3/#natural-dimensions) to [HTML natural sizing definition](https://www.w3.org/TR/css-images-3/#natural-dimensions). Not sure if I'll have time to propose a PR for this myself :(

> 2. The rules for extracting natural dimensions are very specific and cannot be applied to PNG, despite PNG images with resolution units being used for some related examples, eg https://webkit.org/demos/srcset/. These are _false friends_ - the resolution units are in the files, the example is demonstrating resolution units, but those facts are unrelated: the units from the file are not used. This only became clear when we revisited [#6241](https://github.com/w3c/csswg-drafts/issues/6241) and found it meant scaling was now being applied twice.

Why can they not apply to PNG? `PNG` can include `EXIF` tags embedded in `XMP`. In fact, the Mac/iOS Safari implementation does this out of the box because it relies on the Core Foundation libraries that parse XMP-EXIF internally.

**Comment by faceless2** [comment]
Fair point on EXIF in XMP - good to know that's the intention, thanks for letting me know.

I agree a backlink from css-images natural sizing to HTML natural sizing would help significantly, given that's where the algorithm is specified. SVG2 [auto-sizing of images](https://svgwg.org/svg2-draft/embedded.html#Placement) links to css-images so no changes required there.

**Comment by svgeesus** [comment]
>  PNG can include EXIF tags embedded in XMP.

True. PNG also has [an `eXIf` chunk](https://w3c.github.io/PNG-spec/#eXIf)

**Comment by tabatkins** [comment]
> I am surprised though that there is no explicit link from [CSS natural sizing definition](https://www.w3.org/TR/css-images-3/#natural-dimensions) to [HTML natural sizing definition](https://www.w3.org/TR/css-images-3/#natural-dimensions).

Those are both the same link.

**Comment by noamr** [comment]
> > I am surprised though that there is no explicit link from [CSS natural sizing definition](https://www.w3.org/TR/css-images-3/#natural-dimensions) to [HTML natural sizing definition](https://www.w3.org/TR/css-images-3/#natural-dimensions).
> 
> 
> 
> Those are both the same link.

Oops, wrong paste, I meant https://html.spec.whatwg.org/multipage/images.html#density-corrected-intrinsic-width-and-height

But seems like the HTML spec has a reference to setting the CSS natural image dimensions here: https://html.spec.whatwg.org/multipage/images.html#prepare-an-image-for-presentation

Would be good to have a back reference from CSS somehow.

**Comment by tabatkins** [comment]
Yup, would definitely be good, I'll see what I can do. Reason it's lacking currently is the CSS text was written before the HTML stuff was, and I haven't realized I needed to go back and do more edits.

**Comment by AndresInSpace** [comment]
If this is not in same topic of the thread, I apologize, the topic seems to be for discussing issues between both html and css defining images, and those processes / specs.
I believe it's worth mentioning that there is a discussion in HTML wg [#4421](https://github.com/whatwg/html/issues/4421) regarding an issue where at the browser level there is **no optimization happening** for the current selection process of the image src from a provided srcset, which I believe translates to image-set in CSS and seems to go against [CSS spec](https://www.w3.org/TR/css-images-4/#image-set-notation:~:text=Ideally%2C%20images%20should,image%20to%20load.) and [mdn documentation](https://developer.mozilla.org/en-US/docs/Web/CSS/image/image-set#browser_compatibility:~:text=Resolution%20can%20be%20used%20as%20a%20proxy%20for%20filesize%20%E2%80%94%20a%20user%20agent%20on%20a%20slow%20mobile%20connection%20with%20a%20high%2Dresolution%20screen%20may%20prefer%20to%20receive%20lower%2Dresolution%20images%20rather%20than%20waiting%20for%20a%20higher%20resolution%20image%20to%20load) regarding optimization
